src/key.rs

```diff
@@ -1998,21 +2115,21 @@ mod test {
     fn test_tweak_add_assign_then_tweak_add_check() {
         let s = Secp256k1::new();
 
+        // TODO: 10 times is arbitrary, we should test this a _lot_ of times.
         for _ in 0..10 {
-            let mut tweak = [0u8; 32];
-            thread_rng().fill_bytes(&mut tweak);
+            let tweak = random_32_bytes(&mut thread_rng());
 
-            let mut kp = KeyPair::new(&s, &mut thread_rng());
-            let (mut pk, _parity) = kp.x_only_public_key();
+            let kp = KeyPair::new(&s, &mut thread_rng());
+            let (xonly, _parity) = XOnlyPublicKey::from_keypair(&kp);
 
-            let orig_pk = pk;
-            kp.tweak_add_assign(&s, &tweak).expect("Tweak error");
-            let parity = pk.tweak_add_assign(&s, &tweak).expect("Tweak error");
+            let tweaked_kp = kp.add_tweak(&s, &tweak).expect("keypair tweak add failed");
+            let (tweaked_xonly, parity) = xonly.add_tweak(&s, &tweak).expect("xonly pubkey tweak failed");
 
-            let (back, _) = XOnlyPublicKey::from_keypair(&kp);
+            let (want_tweaked_xonly, other_parity) = XOnlyPublicKey::from_keypair(&tweaked_kp);
+            assert_eq!(parity, other_parity); // Sanity check.
```

Comment:
Crypto-geeks, this assertion sometimes fails. Is it correct or do I not understand what is going on here? Thanks

Reply:
See the failing test: https://github.com/rust-bitcoin/rust-secp256k1/runs/6806870441?check_suite_focus=true

Reply:
I haven't fully thought it through but I think it should pass like this:
```suggestion
            assert_eq!(_parity ^ parity, other_parity); // Sanity check.
```

Reply:
Thanks for looking @real-or-random, the suggestion did not work (at least locally). I just remove the whole 'sanity check' and passed the parity returned by `add_tweak` back to `tweak_add_check`. Does this mean there is some bug in the parity handling or that we should _not_ be returning the parity from `XOnlyPublicKey::from_keypair`?

Reply:
```
            // kp = P
            let kp = KeyPair::new(&s, &mut thread_rng());

            // xonly = x(P), i.e., lift_x(x(P)) = ±P
            let (xonly, _) = XOnlyPublicKey::from_keypair(&kp);

            // tweaked_kp = ±P + tG
            let tweaked_kp = kp.add_tweak(&s, &tweak).expect("keypair tweak add failed");

            // tweaked_xonly = x(±P + tG)
            let (tweaked_xonly, parity) = xonly.add_tweak(&s, &tweak).expect("xonly pubkey tweak failed");

            // want_tweaked_xonly = x(±P + tG)
            let (want_tweaked_xonly, tweaked_kp_parity) = XOnlyPublicKey::from_keypair(&tweaked_kp);

            // x(±P + tG) == x(±P + tG)
            assert_eq!(tweaked_xonly, want_tweaked_xonly);

            // lift_x(x(±P + tG)) == lift_x(x(±P + tG))
            assert_eq!(parity, tweaked_kp_parity);
```

So `KeyPair::add_tweak` confused me a lot... It calls `secp256k1_keypair_xonly_tweak_add`, which does two things: It converts the current pubkey to xonly and then tweaks the resulting xonly key. From the name `KeyPair::add_tweak`, I expect it to tweak the keypair directly without converting to xonly first. So maybe this should be called `KeyPair::add_xonly_tweak`. (There's no `secp256k1_keypair_xonly_tweak_add` upstream but this is simply because noone has needed this function so far.)

What adds to the confusion here is that `KeyPair::add_tweak` doesn't output the parity resulting from internal xonly conversation (because `secp256k1_keypair_xonly_tweak_add` doesn't output it either...) This is what should be equal to `_`.

So the way to think about tweaking is that there are basically two modes of tweaking: Ordinary tweaking and x-only tweaking (see https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki#tweaking-definition where we introduce this terminology). This keeps confusing people, including ourselves. We should probably write a blog post on this.

h/t @jonasnick who help me understand what's going on here.